pkg/models/task_duplicate.go

```diff
@@ -0,0 +1,166 @@
+			DueDate:     orig.DueDate,
+			ProjectID:   orig.ProjectID,
+			RepeatAfter: orig.RepeatAfter,
+			RepeatMode:  orig.RepeatMode,
+			Priority:    orig.Priority,
+			StartDate:   orig.StartDate,
+			EndDate:     orig.EndDate,
+			Assignees:   orig.Assignees,
+			Labels:      orig.Labels,
+			HexColor:    orig.HexColor,
+			PercentDone: orig.PercentDone,
+		}
+	}
+
+	// Map from old task ID to new task pointer
+	idMap := map[int64]*Task{}
+
+	// Recursively duplicate a task and its subtasks
+	var duplicateTaskTree func(parentID, origID int64) (*Task, error)
+	duplicateTaskTree = func(parentID, origID int64) (*Task, error) {
+		// Get the original task with all info
+		origTask, err := GetTaskByIDSimple(s, origID)
+		if err != nil {
+			return nil, err
+		}
+		err = addMoreInfoToTasks(s, map[int64]*Task{origTask.ID: &origTask}, doer, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		// Copy the task
+		newTask := copyTask(&origTask)
+		newTask.ProjectID = origTask.ProjectID
+		newTask.ID = 0 // Ensure new insert
+		// Insert the new task
+		err = newTask.Create(s, doer)
+		if err != nil {
+			return nil, err
+		}
+		idMap[origTask.ID] = newTask
+
+		// Duplicate subtasks recursively
+		relations := []*TaskRelation{}
+		err = s.Where("task_id = ? AND relation_kind = ?", origTask.ID, RelationKindSubtask).Find(&relations)
+		if err != nil {
+			return nil, err
+		}
+		for _, rel := range relations {
+			child, err := duplicateTaskTree(newTask.ID, rel.OtherTaskID)
+			if err != nil {
+				return nil, err
+			}
+			// Create subtask relation (newTask.ID -> child.ID)
+			tr := &TaskRelation{
+				TaskID:       newTask.ID,
+				OtherTaskID:  child.ID,
+				RelationKind: RelationKindSubtask,
+			}
+			if err := tr.Create(s, doer); err != nil {
+				return nil, err
+			}
+		}
+		return newTask, nil
+	}
+
+	// Start duplicating from the root task
+	root, err := duplicateTaskTree(0, td.TaskID)
+	if err != nil {
+		return err
+	}
+	td.Task = root
+
+	// Copy follows/precedes relations for all duplicated tasks, needs to select only one type, corresponding will be created too
+	for oldID, newTask := range idMap {
+		rels := []*TaskRelation{}
+		err := s.Where("task_id = ? AND relation_kind = ? ", oldID, RelationKindFollows).Find(&rels)
+		if err != nil {
+			return err
+		}
+		for _, rel := range rels {
+			// Only copy if the other task is also duplicated
+			if other, ok := idMap[rel.OtherTaskID]; ok {
+				tr := &TaskRelation{
+					TaskID:       newTask.ID,
+					OtherTaskID:  other.ID,
+					RelationKind: rel.RelationKind,
+				}
+				if err := tr.Create(s, doer); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// CanCreate checks if a user has the right to duplicate a task
+func (td *TaskDuplicate) CanCreate(s *xorm.Session, a web.Auth) (canCreate bool, err error) {
+	// project Exists + user has read and write access to project
+	project := &Project{ID: td.ProjectID}
+	canRead, _, err := project.CanRead(s, a)
```

Comment:
If the user has write access to a project, they can also read it, no need to check read access separately.